### PR TITLE
Data source bug fixes

### LINKED ISSUE #2533: National Data->Data Sources  errors
- [x] selecting same value breaks selection (this is probably caused by an inconsistency in auto complete component onChange must return same object type {value, label})
- [x] deleting a row leaves values in placeholder inputs
- [ ] migration fails when there are no data sources to migrate 

https://user-images.githubusercontent.com/1138156/231080279-45d15a9b-746b-4045-b4e5-a4551e2481ac.mov

- [ ] add prefixes for variable names in table 2b

### PR DESCRIPTION
**Pull request by sorja**
- [x] fix:  selecting same value breaks selection (this is probably caused by an inconsistency in auto complete component onChange must return same object type {value, label})
- [x]  deleting a row leaves values in placeholder inputs ( fixed by @rfontanarosa )
- [x] migration fails when there are no data sources to migrate
- [x] datasource prefix labels

image

### CODE REVIEW

File: src/client/components/Autocomplete/Autocomplete.tsx [hunk 1]
```diff
@@ -50,6 +50,18 @@ const Autocomplete: React.FC<Props> = (props: Props) => {
...
+      // Disable saving if selected item is not changed when free text disabled
+      if (withArrow && !changes.selectedItem) {
+        return
+      }
+
+      // When arrow is shown, save only selected item
+      if (withArrow) {
+        onSave(changes.selectedItem ?? '')
+        return
+      }
```
Comment (sorja) [inline src/client/components/Autocomplete/Autocomplete.tsx]: > selecting same value breaks selection (this is probably caused by an inconsistency in auto complete component onChange must return same object type {value, label})

File: src/client/pages/AssessmentSection/Descriptions/Description/DataSources/DataSourceColumn/ColumnFraVariable.tsx [hunk 2]
```diff
@@ -27,11 +27,18 @@ const ColumnFraVariable: React.FC<Props> = (props: Props) => {
...
+      if (dataSourceVariables?.prefixes?.[row.props.variableName]) {
+        prefix = Labels.getLabel({ label: dataSourceVariables?.prefixes?.[row.props.variableName], t })
+        prefix += ' '
+      }
+
       const _option = {
-        label: t(Cols.getLabel({ cycle, col: row.cols[0], t })),
+        label: `${prefix}${label}`,
         value: row.props.variableName,
       }
```
Comment (sorja) [inline src/client/pages/AssessmentSection/Descriptions/Description/DataSources/DataSourceColumn/ColumnFraVariable.tsx:40]: datasource prefix labels

File: src/test/migrations/steps/20230403165717-step-map-datasource-freetext-to-constant.ts [hunk 3]
```diff
@@ -54,6 +54,8 @@ export default async (client: BaseProtocol) => {
     (row) => Objects.camelize(row)
   )
 
+  if (descriptions.length < 1) return
```
Comment (sorja) [inline src/test/migrations/steps/20230403165717-step-map-datasource-freetext-to-constant.ts:57]: > migration fails when there are no data sources to migrate

File: src/test/migrations/steps/20230411163053-step-2b-data-source-prefixes.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+  const ranks = [...nativeRanks, ...introducedRanks].join(',\n')
+
+  const query = `
+    update ${schemaName}.section s
+    set props = jsonb_set(
+        props::jsonb,
+        '{descriptions,${cycle.uuid},nationalData,dataSources,table,dataSourceVariables,prefixes}',
+        '{${ranks}}'
+    )::jsonb
+    where s.props ->> 'name' = 'growingStockComposition'
```
Comment (sorja) [inline src/test/migrations/steps/20230411163053-step-2b-data-source-prefixes.ts:25]: > datasource prefix labels

File: src/meta/assessment/col.ts [hunk 5]
```diff
@@ -1,6 +1,8 @@
 import { CycledPropsObject, CycleUuid } from './cycle'
 import { Label } from './label'
 
+export type ColName = string
```
Comment (minotogna) [inline src/meta/assessment/col.ts:4]: export this from '@meta/assessment' as well

File: src/client/components/Autocomplete/Autocomplete.tsx (hunk 1 above)
Comment (minotogna) [reply src/client/components/Autocomplete/Autocomplete.tsx]: I think withArrow is not meaningful . can we rename the prop ?

**Review by minotogna — APPROVED** [review]

**Review by rfontanarosa — APPROVED** [review]

**Pull request merged**